### PR TITLE
4.x features

### PR DESCRIPTION
**Pull request by tomas-langer**
New feature API, processor and catalog.

See docs-internal/feature-catalog.md for details about changes.

- introduces `@Aot` annotation to define whether native image is supported (or to add description with limitations)
- introduces `@Preview` for features that are to stay with Helidon, but are in the progress of shaping their APIs
- introduces `@Incubating` for features that may be removed, or heavily modified
- introduces `@Feature` as the feature definition

All of these annotations are to be used on `module-info.java` module definition

`@Preview` features - the first implementation of reactive messaging was a preview feature - we knew we will support that specification, but wanted to have some time to make sure everything is shaped correctly
`@Incubating` features - our custom reactive OCI  integration was incubating (and later removed) - we were not sure it is the right thing to do, wanted to try it out, decided to discontinue...

### CODE REVIEW

**Comment by m0mus** [comment]
What's the difference between `@Incubating` and `@Experimental`. I am not convinced that we need `@Experimental` as it's described. Can you provide some samples?

**Review by barchetta — CHANGES_REQUESTED** [review]

File: common/features/api/src/main/java/io/helidon/common/features/api/Experimental.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+package io.helidon.common.features.api;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Annotation for experimental modules.
+ * Experimental module may be changed including backward incompatible changes in between major versions of Helidon.
```
[the remark this answers is not in the document]
Comment (tomas-langer) [reply common/features/api/src/main/java/io/helidon/common/features/api/Experimental.java:26]: So have `Preview` and `Experimental`, or `Preview` and `Incubating`? 
I would try to keep only two of these

[the remark this answers is not in the document]
Comment (barchetta) [reply common/features/api/src/main/java/io/helidon/common/features/api/Experimental.java:26]: I am proposing "Preview" and "Incubating".

[the remark this answers is not in the document]
Comment (tomas-langer) [reply common/features/api/src/main/java/io/helidon/common/features/api/Experimental.java:26]: OK, I will refactor the PR

**Comment by tomas-langer** [comment]
Rebased to resolve conflict

**Comment by tomas-langer** [comment]
> What's the difference between `@Incubating` and `@Experimental`. I am not convinced that we need `@Experimental` as it's described. Can you provide some samples?

Updated description of PR with examples

File: common/features/api/src/main/java/io/helidon/common/features/api/Incubating.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+package io.helidon.common.features.api;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Annotation for incubating feature modules.
+ * Incubating features may be changed including backward incompatible changes in between major versions of Helidon.
```
Comment (barchetta) [inline common/features/api/src/main/java/io/helidon/common/features/api/Incubating.java:26]: "between major versions of Helidon" -> "between any version of Helidon"

File: common/features/api/src/main/java/io/helidon/common/features/api/Preview.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+package io.helidon.common.features.api;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Annotation for preview feature modules.
+ * Preview feature may be changed including backward incompatible changes in between major versions of Helidon.
```
Comment (barchetta) [inline common/features/api/src/main/java/io/helidon/common/features/api/Preview.java:26]: Maybe change "major" to "minor"? We allow API changes for any API between major versions.  I think we should have a bit more flexibility with Preview APIs.

**Review by barchetta — APPROVED** [review]

**Pull request merged**